File: server/app/schemas/personas.py
"""Personas V2 API schemas."""

from pydantic import BaseModel

from .base import (DepartmentMapping, ModelMapping, ReasoningMapping,
                   ScenarioMapping)


class PersonasFilters(BaseModel):
    """Filters for personas list request."""

    departmentIds: list[str]
    profileId: str


class PersonaItem(BaseModel):
    """Individual persona item in the response."""

    persona_id: str
    name: str  # Added name
    description: str | None
    color: str
    icon: str
    department_ids: list[str] | None  # None = cross-department (all departments)
    scenario_ids: list[str]  # Array of scenario IDs
    model_id: str
    reasoning: str | None
    temperature: float
    active: bool
    num_scenarios: int
    can_edit: bool
    can_duplicate: bool
    can_delete: bool


class PersonasListResponse(BaseModel):
    """Response for personas list endpoint."""

    personas: list[PersonaItem]
    scenario_mapping: ScenarioMapping
    model_mapping: ModelMapping


class DuplicatePersonaRequest(BaseModel):
    """Request to duplicate a persona."""

    personaId: str


class DuplicatePersonaResponse(BaseModel):
    """Response from duplicate operation."""

    success: bool
    personaId: str
    message: str


class DeletePersonaRequest(BaseModel):
    """Request to delete a persona."""

    personaId: str


class DeletePersonaResponse(BaseModel):
    """Response from delete operation."""

    success: bool
    message: str


class PersonaDetailRequest(BaseModel):
    """Request to get persona details."""

    personaId: str
    profileId: str


class DebugInfoItem(BaseModel):
    """Debug information item."""

    created_at: str
    model_id: str
    content: str


class PromptInfo(BaseModel):
    """Prompt information for version history."""

    system_prompt: str
    created_at: str
    updated_at: str
    department_ids: list[str] | None


class PersonaDetailResponse(BaseModel):
    """Detailed persona response with all fields and metadata."""

    # Basic persona fields
    name: str
    description: str | None
    department_ids: list[str] | None  # None = cross-department (all departments)
    active: bool
    color: str
    icon: str
    model_id: str
    reasoning: str | None
    temperature: float
    system_prompt: str
    prompt_id: str | None

    # Usage and permissions
    in_use: bool
    scenario_count: int
    can_edit: bool
    can_duplicate: bool
    can_delete: bool

    # Metadata/Options
    preset_colors: list[str]
    suggested_icons: list[str]
    valid_icons: list[str]
    valid_model_ids: list[str]
    reasoning_options: list[str]
    valid_department_ids: list[str]
    temperature_lower: float
    temperature_upper: float

    # Prompt version history
    prompt_mapping: dict[str, PromptInfo]

    # Mappings
    model_mapping: ModelMapping
    reasoning_mapping: ReasoningMapping
    department_mapping: DepartmentMapping

    # Debug info
    debug_info: list[DebugInfoItem]


class PersonaDetailDefaultRequest(BaseModel):
    """Request to get default persona details."""

    profileId: str


# Response uses same PersonaDetailResponse


class CreatePersonaRequest(BaseModel):
    """Request to create a persona."""

    name: str
    description: str | None
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    color: str
    icon: str
    model_id: str
    reasoning: str | None
    temperature: float
    prompt_id: str | None  # If provided, use existing prompt
    system_prompt: str  # If prompt_id is None, create new prompt with this


class CreatePersonaResponse(BaseModel):
    """Response from create operation."""

    success: bool
    personaId: str
    message: str


class UpdatePersonaRequest(BaseModel):
    """Request to update a persona."""

    personaId: str
    name: str
    description: str | None
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    color: str
    icon: str
    model_id: str
    reasoning: str | None
    temperature: float
    prompt_id: str | None  # If provided, use existing prompt
    system_prompt: str  # If prompt_id is None, create new prompt with this


class UpdatePersonaResponse(BaseModel):
    """Response from update operation."""

    success: bool
    message: str

File: client/lib/api/v2/schemas/personas.ts
/**
 * Personas V2 API Schemas
 * Schema definitions for personas v2 endpoints
 */

import { z } from "zod";
import {
  DepartmentMappingSchema,
  ModelMappingSchema,
  ReasoningMappingSchema,
  ScenarioMappingSchema,
} from "./base";

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const PersonasFiltersSchema = z.object({
  profileId: z.string(),
});

export type PersonasFilters = z.infer<typeof PersonasFiltersSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Persona item (normalized - no nested objects, just IDs)
export const PersonaItemSchema = z.object({
  persona_id: z.string(),
  name: z.string(), // Added name back
  description: z.string().nullable(),
  color: z.string(),
  icon: z.string(),
  department_ids: z.array(z.string()).nullable(), // None = cross-department (all departments)
  scenario_ids: z.array(z.string()), // Array of scenario IDs
  model_id: z.string(),
  reasoning: z.string().nullable(),
  temperature: z.number(),
  active: z.boolean(),
  num_scenarios: z.number(),
  can_edit: z.boolean(),
  can_duplicate: z.boolean(),
  can_delete: z.boolean(),
});

// Response Schema (normalized with top-level mappings, no persona_mapping)
export const PersonasListResponseSchema = z.object({
  personas: z.array(PersonaItemSchema),
  scenario_mapping: ScenarioMappingSchema,
  model_mapping: ModelMappingSchema,
  department_mapping: DepartmentMappingSchema,
});

export type PersonasListResponse = z.infer<typeof PersonasListResponseSchema>;
export type PersonaItem = z.infer<typeof PersonaItemSchema>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Duplicate request
export const DuplicatePersonaRequestSchema = z.object({
  personaId: z.string(),
});

export type DuplicatePersonaRequest = z.infer<
  typeof DuplicatePersonaRequestSchema
>;

// Duplicate response
export const DuplicatePersonaResponseSchema = z.object({
  success: z.boolean(),
  personaId: z.string(),
  message: z.string(),
});

export type DuplicatePersonaResponse = z.infer<
  typeof DuplicatePersonaResponseSchema
>;

// Delete request
export const DeletePersonaRequestSchema = z.object({
  personaId: z.string(),
});

export type DeletePersonaRequest = z.infer<typeof DeletePersonaRequestSchema>;

// Delete response
export const DeletePersonaResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeletePersonaResponse = z.infer<typeof DeletePersonaResponseSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

// Detail request
export const PersonaDetailRequestSchema = z.object({
  personaId: z.string(),
  profileId: z.string(),
});

export type PersonaDetailRequest = z.infer<typeof PersonaDetailRequestSchema>;

// Debug info item
export const DebugInfoItemSchema = z.object({
  created_at: z.string(),
  model_id: z.string(),
  content: z.string(),
});

export type DebugInfoItem = z.infer<typeof DebugInfoItemSchema>;

// Detail response
export const PersonaDetailResponseSchema = z.object({
  // Basic persona fields
  name: z.string(),
  description: z.string().nullable(),
  department_ids: z.array(z.string()).nullable(),
  active: z.boolean(),
  color: z.string(),
  icon: z.string(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  temperature: z.number(),
  system_prompt: z.string(),
  prompt_id: z.string().nullable(),

  // Prompt version history
  prompt_mapping: z.record(
    z.string(),
    z.object({
      system_prompt: z.string(),
      created_at: z.string(),
      updated_at: z.string(),
      department_ids: z.array(z.string()).nullable(),
    })
  ),

  // Usage and permissions
  in_use: z.boolean(),
  scenario_count: z.number(),
  can_edit: z.boolean(),
  can_duplicate: z.boolean(),
  can_delete: z.boolean(),

  // Metadata/Options
  preset_colors: z.array(z.string()),
  suggested_icons: z.array(z.string()),
  valid_icons: z.array(z.string()),
  valid_model_ids: z.array(z.string()),
  reasoning_options: z.array(z.string()),
  valid_department_ids: z.array(z.string()),
  temperature_lower: z.number(),
  temperature_upper: z.number(),

  // Mappings (using centralized types)
  model_mapping: ModelMappingSchema,
  reasoning_mapping: ReasoningMappingSchema,
  department_mapping: DepartmentMappingSchema,

  // Debug info
  debug_info: z.array(DebugInfoItemSchema),
});

export type PersonaDetailResponse = z.infer<typeof PersonaDetailResponseSchema>;

// Default detail request (profile-based, no persona_id)
export const PersonaDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type PersonaDetailDefaultRequest = z.infer<
  typeof PersonaDetailDefaultRequestSchema
>;

// Response uses same PersonaDetailResponse schema

// ============================================================================
// CREATE SCHEMAS
// ============================================================================

// Create request
export const CreatePersonaRequestSchema = z.object({
  name: z.string(),
  description: z.string().nullable(),
  department_ids: z.array(z.string()).nullable(),
  active: z.boolean(),
  color: z.string(),
  icon: z.string(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  temperature: z.number(),
  prompt_id: z.string().nullable().optional(), // If provided, use existing prompt
  system_prompt: z.string(), // If prompt_id is None, create new prompt with this
});

export type CreatePersonaRequest = z.infer<typeof CreatePersonaRequestSchema>;

// Create response
export const CreatePersonaResponseSchema = z.object({
  success: z.boolean(),
  personaId: z.string(),
  message: z.string(),
});

export type CreatePersonaResponse = z.infer<typeof CreatePersonaResponseSchema>;

// Update request (same fields as create)
export const UpdatePersonaRequestSchema = z.object({
  personaId: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  department_ids: z.array(z.string()).nullable(),
  active: z.boolean(),
  color: z.string(),
  icon: z.string(),
  model_id: z.string(),
  reasoning: z.string().nullable(),
  temperature: z.number(),
  prompt_id: z.string().nullable().optional(), // If provided, use existing prompt
  system_prompt: z.string(), // If prompt_id is None, create new prompt with this
});

export type UpdatePersonaRequest = z.infer<typeof UpdatePersonaRequestSchema>;

export const UpdatePersonaResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type UpdatePersonaResponse = z.infer<typeof UpdatePersonaResponseSchema>;
